android/app/src/test/java/com/threed/supply/LoginLogicTest.kt

```diff
@@ -0,0 +1,165 @@
+package com.threed.supply
+
+import com.spotify.mobius.test.NextMatchers.hasEffects
+import com.spotify.mobius.test.NextMatchers.hasModel
+import com.spotify.mobius.test.UpdateSpec
+import com.spotify.mobius.test.UpdateSpec.assertThatNext
+import com.threed.supply.features.login.domain.LoginEffect
+import com.threed.supply.features.login.domain.LoginEffect.*
+import com.threed.supply.features.login.domain.LoginEvent
+import com.threed.supply.features.login.domain.LoginEvent.*
+import com.threed.supply.features.login.domain.LoginModel
+import com.threed.supply.features.login.domain.update
+import org.junit.Before
+import org.junit.Test
+
+class LoginLogicTest {
+
+    private lateinit var updateSpec: UpdateSpec<LoginModel, LoginEvent, LoginEffect>
+
+    @Before
+    fun setUp() {
+        updateSpec = UpdateSpec(::update)
+    }
+
+    @Test
+    fun updateEmail_whenEmailChanged() {
+        val model = LoginModel()
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test"))
+            .then(assertThatNext(hasModel(model.copy(email = "test"))))
+    }
+
+    @Test
+    fun updatePassword_whenPasswordChanged() {
+        val model = LoginModel()
+        updateSpec
+            .given(model)
+            .whenEvent(PasswordChanged(password = "1234"))
+            .then(assertThatNext(hasModel(model.copy(password = "1234"))))
+    }
+
+    @Test
+    fun enableLogin_whenEmailChanged_andValidEmailAndPassword() {
+        val model = LoginModel(password = "12345678")
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test@gmail.com"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(email = "test@gmail.com", loginEnabled = true))
+                )
+            )
+    }
+
+    @Test
+    fun enableLogin_whenPasswordChanged_andValidEmailAndPassword() {
+        val model = LoginModel(email = "test@gmail.com")
+        updateSpec
+            .given(model)
+            .whenEvent(PasswordChanged(password = "12345678"))
+            .then(
+                assertThatNext(
+                    hasModel(model.copy(password = "12345678", loginEnabled = true))
+                )
+            )
+    }
+
+    @Test
+    fun disableLogin_whenInvalidEmail() {
+        val model = LoginModel(email = "test@gmail.com", loginEnabled = true)
+        updateSpec
+            .given(model)
+            .whenEvent(EmailChanged(email = "test@gmail.com"))
```

Comment:
why is this invalid? Is it because it's the same as the original one? Why is test not failing?